From: Michael Holzheu <holzheu@linux.vnet.ibm.com>
Subject: Revert "s390/kdump: Clear subchannel ID to signal non-CCW/SCSI IPL"
Patch-mainline: v4.7-rc6
Git-commit: 5419447e2142d6ed68c9f5c1a28630b3a290a845
References: bnc#988934, LTC#143668

Description:  kernel: Revert clearing of subchannel ID for non-CCW/SCSI IPL
Symptom:      When using Linux as boot kernel that loads another kernel via
              kexec the lsreipl tool shows "Re-IPL type:: unknown (unknown)".
Problem:      Boot information of old kernel is cleared via kexec. This
              was introduced by upstream commit 852ffd0f4e23248b4753
              "s390/kdump: Clear subchannel ID to signal non-CCW/SCSI IPL".
Solution:     Revert upstream commit.
Reproduction: 1) Boot first Linux
              2) Load next Linux with:
                 # kexec -l /boot/vmlinux --initrd=/boot/initrd \
                   --append=$(cat /proc/cmdline)
              3) Trigger reboot with:
                 # kexec -e
              4) Check lsreipl output:
                 # lsreipl
                   Re-IPL type:: unknown (unknown)

Upstream-Description:

              Revert "s390/kdump: Clear subchannel ID to signal non-CCW/SCSI IPL"

              This reverts commit 852ffd0f4e23248b47531058e531066a988434b5.

              There are use cases where an intermediate boot kernel (1) uses kexec
              to boot the final production kernel (2). For this scenario we should
              provide the original boot information to the production kernel (2).
              Therefore clearing the boot information during kexec() should not
              be done.

              Cc: stable@vger.kernel.org # v3.17+
              Reported-by: Steffen Maier <maier@linux.vnet.ibm.com>
              Signed-off-by: Michael Holzheu <holzheu@linux.vnet.ibm.com>
              Reviewed-by: Heiko Carstens <heiko.carstens@de.ibm.com>
              Signed-off-by: Martin Schwidefsky <schwidefsky@de.ibm.com>


Signed-off-by: Michael Holzheu <holzheu@linux.vnet.ibm.com>
Acked-by: John Jolly <jjolly@suse.de>
---
 arch/s390/kernel/ipl.c |    7 -------
 1 file changed, 7 deletions(-)

--- a/arch/s390/kernel/ipl.c
+++ b/arch/s390/kernel/ipl.c
@@ -2070,13 +2070,6 @@ void s390_reset_system(void (*fn_pre)(vo
 	S390_lowcore.program_new_psw.addr =
 		PSW_ADDR_AMODE | (unsigned long) s390_base_pgm_handler;
 
-	/*
-	 * Clear subchannel ID and number to signal new kernel that no CCW or
-	 * SCSI IPL has been done (for kexec and kdump)
-	 */
-	S390_lowcore.subchannel_id = 0;
-	S390_lowcore.subchannel_nr = 0;
-
 	/* Store status at absolute zero */
 	store_status();
 
